pkg/puller/puller.go

```diff
@@ -101,51 +101,59 @@ func (p *Puller) manage(ctx context.Context, warmupTime time.Duration) {
 
 	var prevRadius uint8
 
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-c:
+	onChange := func() {
+		p.syncPeersMtx.Lock()
 
-			p.syncPeersMtx.Lock()
+		// peersDisconnected is used to mark and prune peers that are no longer connected.
+		peersDisconnected := make(map[string]*syncPeer)
+		for _, peer := range p.syncPeers {
+			peersDisconnected[peer.address.ByteString()] = peer
+		}
 
-			// peersDisconnected is used to mark and prune peers that are no longer connected.
-			peersDisconnected := make(map[string]*syncPeer)
-			for _, peer := range p.syncPeers {
-				peersDisconnected[peer.address.ByteString()] = peer
-			}
+		neighborhoodDepth := p.topology.NeighborhoodDepth()
+		syncRadius := p.reserveState.GetReserveState().StorageRadius
 
-			neighborhoodDepth := p.topology.NeighborhoodDepth()
-			syncRadius := p.reserveState.GetReserveState().StorageRadius
+		// if the radius decreases, we must fully resync the bin
+		if syncRadius < prevRadius {
+			err := p.resetInterval(syncRadius)
+			if err != nil {
+				p.logger.Error(err, "reset lower sync radius")
+			}
+		}
+		prevRadius = syncRadius
 
-			// if the radius decreases, we must fully resync the bin
-			if syncRadius < prevRadius {
-				err := p.resetInterval(syncRadius)
-				if err != nil {
-					p.logger.Error(err, "reset lower sync radius")
+		_ = p.topology.EachPeerRev(func(addr swarm.Address, po uint8) (stop, jumpToNext bool, err error) {
+			if po >= neighborhoodDepth {
+				// add peer to sync
+				if _, ok := p.syncPeers[addr.ByteString()]; !ok {
+					p.syncPeers[addr.ByteString()] = newSyncPeer(addr, p.bins)
 				}
+				// remove from disconnected list as the peer is still connected
+				delete(peersDisconnected, addr.ByteString())
 			}
-			prevRadius = syncRadius
-
-			_ = p.topology.EachPeerRev(func(addr swarm.Address, po uint8) (stop, jumpToNext bool, err error) {
-				if po >= neighborhoodDepth {
-					// add peer to sync
-					if _, ok := p.syncPeers[addr.ByteString()]; !ok {
-						p.syncPeers[addr.ByteString()] = newSyncPeer(addr, p.bins)
-					}
-					// remove from disconnected list as the peer is still connected
-					delete(peersDisconnected, addr.ByteString())
-				}
-				return false, false, nil
-			}, topology.Filter{Reachable: true})
+			return false, false, nil
+		}, topology.Filter{Reachable: true})
 
-			for _, peer := range peersDisconnected {
-				p.disconnectPeer(peer.address)
-			}
+		for _, peer := range peersDisconnected {
+			p.disconnectPeer(peer.address)
+		}
+
+		p.recalcPeers(ctx, syncRadius)
+
+		p.syncPeersMtx.Unlock()
+	}
 
-			p.recalcPeers(ctx, syncRadius)
+	tick := time.NewTicker(5 * time.Minute)
+	defer tick.Stop()
 
-			p.syncPeersMtx.Unlock()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-tick.C:
+			onChange()
+		case <-c:
```

Comment:
So in order to prevent two simultaneous runs of this, would it be better to do a ticker reset here? this should prevent abuse of this I think.